manual/src/tutorials/gadtexamples.etex

```diff
@@ -304,8 +304,149 @@ end
+
+type one = TypeNat.(zero succ)
+\end{caml_example*}
+
+This definition does not work: because "zero" and "succ" are abstract
+types, the compiler assumes that their definition is unknown, and that
+they might in fact be the same type. A module with the same interface
+\emph{could} have been defined as follows:
+\begin{caml_example*}{verbatim}
+module OtherTypeNat : sig
+  type zero
+  type !'a succ  
+end = struct
+  type 'a succ = Foo
+  type zero = unit succ
+end
+\end{caml_example*}
+
+In particular, the types "zero" and "zero succ" are not known to be
+distinct, and the compiler cannot agree that the "Nil" case is
+impossible at type "zero succ":\footnote{We also needed an explicit
+  injectivity annotation, "type !'a succ", otherwise the definition of
+  "sized_list" is also rejected because "succ" is abstract.}
+\begin{caml_example}{verbatim}[error]
+open OtherTypeNat
+let of_singleton : ('a, zero succ) sized_list -> 'a = function
+  | Nil -> .
+  | Cons (v, Nil) -> v
+  | Cons (v, Cons _) -> .
+\end{caml_example}
+
+This problem also affects datatype definitions with the same
+constructor or field names, which could be alias of each other.
+\begin{caml_example*}{verbatim}[error]
+module DataTypeNat = struct
+  type zero = TypeNat
+  type 'a succ = TypeNat
+end
+open DataTypeNat
+
+let of_singleton : ('a, zero succ) sized_list -> 'a = function
+  | Nil -> .
+  | Cons (v, Nil) -> v
+  | Cons (v, Cons _) -> .
+\end{caml_example*}
+
+The recommended way to avoid this problem is to give a concrete
+definition to types such as "zero" and "succ", that are only meant to
+be used in GADT indices. (This is counter-intuitive, as we will never
+manipulate values of this type.) You could use:
+\begin{caml_example*}{verbatim}
+module TypeNat = struct
+  type zero = Zero
+  type 'a succ = Succ
+end
+\end{caml_example*}
+The types are now incompatibles as their constructor names are distinct.
+
+As an additional refinement, we can define those index types using
+using a "private" annotation, which clarifies that they are not meant
+to be used to build values:
+\begin{caml_example*}{verbatim}
+module TypeNat = struct
+  type zero = private Zero
+  type 'a succ = private Succ
+end
+\end{caml_example*}
+
+Remark: Usually "private" is meant to be used in interfaces, not in
+implementations; see the documentation of private types in
+\ref{s:private-types}. The constructors of a datatype that is
+\emph{defined} as "private" cannot be used to build values, so the
+type is effectively empty. This makes it more explicit that the types
+are only intended to be used as GADT indices.
+
+\subsection{s:abstract-types-current-module}{Legacy behavior: types in the current module}
+
+Early versions of OCaml, up to OCaml 5.3, had a special handling of
+types defined in the current module. (This special case does
+not apply to "succ" and "zero" above, as they are defined in
+a separate "TypeNat" submodule.) So the following definitions used to
```

Comment:
I would move the sentence in parentheses to the "This special case made ..." paragraph since this is an illustration of the modularity issue.